// clude-bot/dist/utils/constants.js
"use strict";
/**
 * Centralized constants — magic numbers, well-known addresses, and configuration defaults.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.KNOWN_MEMECOINS = exports.MAJOR_TOKENS = exports.MEMO_MAX_LENGTH = exports.ERC20_BALANCE_ABI = exports.BASE_WETH_ADDRESS = exports.ALLIUM_BASE_URL = exports.DEXSCREENER_PRICE_URL = exports.BASESCAN_API_BASE_URL = exports.BASESCAN_TX_BASE_URL = exports.BASE_CHAIN_ID = exports.SIDEWAYS_THRESHOLD_PERCENT = exports.PUMP_DUMP_THRESHOLD_PERCENT = exports.WHALE_SELL_COOLDOWN_MS = exports.PRICE_SNAPSHOT_RETENTION_HOURS = exports.REFLECTION_MIN_INTERVAL_MS = exports.REFLECTION_IMPORTANCE_THRESHOLD = exports.EMBEDDING_FRAGMENT_MAX_LENGTH = exports.EMBEDDING_DIMENSIONS = exports.LINK_CO_RETRIEVAL_BOOST = exports.MAX_AUTO_LINKS = exports.LINK_SIMILARITY_THRESHOLD = exports.RETRIEVAL_WEIGHT_GRAPH = exports.MEMORY_CONCEPTS = exports.RETRIEVAL_WEIGHT_VECTOR = exports.RETRIEVAL_WEIGHT_IMPORTANCE = exports.RETRIEVAL_WEIGHT_RELEVANCE = exports.RETRIEVAL_WEIGHT_RECENCY = exports.RECENCY_DECAY_BASE = exports.DECAY_RATES = exports.MEMORY_MAX_SUMMARY_LENGTH = exports.MEMORY_MAX_CONTENT_LENGTH = exports.MEMORY_MIN_DECAY = exports.MEMORY_DECAY_RATE = exports.TWEET_SAFE_LENGTH = exports.TWEET_MAX_LENGTH = exports.SOLSCAN_TX_BASE_URL = exports.HELIUS_BALANCES_BASE_URL = exports.HELIUS_RPC_BASE_URL = exports.JUPITER_PRICE_URL = exports.MEMO_PROGRAM_ID = void 0;
exports.isNoteworthyToken = isNoteworthyToken;
// Solana
exports.MEMO_PROGRAM_ID = 'MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr';
exports.JUPITER_PRICE_URL = 'https://api.jup.ag/price/v2';
exports.HELIUS_RPC_BASE_URL = 'https://mainnet.helius-rpc.com';
exports.HELIUS_BALANCES_BASE_URL = 'https://api.helius.xyz/v0/addresses';
exports.SOLSCAN_TX_BASE_URL = 'https://solscan.io/tx';
exports.TWEET_MAX_LENGTH = 280;
exports.TWEET_SAFE_LENGTH = 270;
// Memory system
exports.MEMORY_DECAY_RATE = 0.95; // Legacy: uniform decay (now per-type below)
exports.MEMORY_MIN_DECAY = 0.05;
exports.MEMORY_MAX_CONTENT_LENGTH = 5000;
exports.MEMORY_MAX_SUMMARY_LENGTH = 500;
// Type-specific decay rates (per 24h cycle)
// Episodic events fade fastest; identity/knowledge persists longest.
exports.DECAY_RATES = {
    episodic: 0.93, // Individual events fade quickly
    semantic: 0.98, // Learned knowledge persists
    procedural: 0.97, // Behavioral patterns are stable
    self_model: 0.99, // Identity is nearly permanent
};
// Memory retrieval — additive scoring (Park et al. 2023, Generative Agents)
exports.RECENCY_DECAY_BASE = 0.995; // Exponential: 0.995^hours since last access
exports.RETRIEVAL_WEIGHT_RECENCY = 0.5; // Paper: recency has lowest weight
exports.RETRIEVAL_WEIGHT_RELEVANCE = 3.0; // Paper: relevance dominates
exports.RETRIEVAL_WEIGHT_IMPORTANCE = 2.0; // Paper: importance is second
exports.RETRIEVAL_WEIGHT_VECTOR = 3.0; // Vector similarity (high but not dominant)
// Structured concept ontology for cross-cutting knowledge classification
// Replaces freeform tagging with a controlled vocabulary for precise retrieval.
exports.MEMORY_CONCEPTS = [
    'market_event', // Price movements, ATH, dumps, volume spikes
    'holder_behavior', // Whale moves, exits, accumulation, selling patterns
    'self_insight', // Self-observations, identity reflections
    'social_interaction', // Mentions, replies, conversations on X
    'community_pattern', // Recurring community behaviors, engagement trends
    'token_economics', // Swaps, liquidity, transfers, on-chain activity
    'sentiment_shift', // Mood changes, market sentiment transitions
    'recurring_user', // Returning users, first interactions, relationship building
    'whale_activity', // Large holder movements, big trades
    'price_action', // Chart patterns, price analysis, technical signals
    'engagement_pattern', // What content resonates, what falls flat
    'identity_evolution', // How Clude is changing, personality shifts
];
// Memory association graph
exports.RETRIEVAL_WEIGHT_GRAPH = 1.5; // Graph-linked memory boost in scoring
exports.LINK_SIMILARITY_THRESHOLD = 0.6; // Min vector similarity for auto-linking
exports.MAX_AUTO_LINKS = 5; // Max links created per new memory
exports.LINK_CO_RETRIEVAL_BOOST = 0.05; // Hebbian strength boost on co-retrieval
// Embedding system
exports.EMBEDDING_DIMENSIONS = 1024;
exports.EMBEDDING_FRAGMENT_MAX_LENGTH = 2000; // Max chars per fragment for granular decomposition
// Event-driven reflection triggers
exports.REFLECTION_IMPORTANCE_THRESHOLD = 2.0; // Cumulative importance to trigger reflection (3-4 interactions)
exports.REFLECTION_MIN_INTERVAL_MS = 30 * 60 * 1000; // Min 30 min between reflections
// Price oracle
exports.PRICE_SNAPSHOT_RETENTION_HOURS = 48;
exports.WHALE_SELL_COOLDOWN_MS = 30 * 60 * 1000;
exports.PUMP_DUMP_THRESHOLD_PERCENT = 10;
exports.SIDEWAYS_THRESHOLD_PERCENT = 2;
// Base chain (dormant — kept for future switch-back)
exports.BASE_CHAIN_ID = 8453;
exports.BASESCAN_TX_BASE_URL = 'https://basescan.org/tx';
exports.BASESCAN_API_BASE_URL = 'https://api.basescan.org/api';
exports.DEXSCREENER_PRICE_URL = 'https://api.dexscreener.com/latest/dex/tokens';
// Allium API
exports.ALLIUM_BASE_URL = 'https://api.allium.so/api/v1/developer';
// WETH on Base (dormant — kept for future switch-back)
exports.BASE_WETH_ADDRESS = '0x4200000000000000000000000000000000000006';
// Minimal ERC-20 ABI for balance queries
exports.ERC20_BALANCE_ABI = [
    'function balanceOf(address) view returns (uint256)',
    'function decimals() view returns (uint8)',
    'function symbol() view returns (string)',
];
// Memo max length for on-chain calldata
exports.MEMO_MAX_LENGTH = 566;
// Token lists for market monitoring
exports.MAJOR_TOKENS = new Set([
    'ETH', 'BTC', 'USDC', 'AERO', 'cbBTC', 'VIRTUAL', 'MORPHO', 'RSR',
    'SOL', 'BONK', 'WIF', 'JTO', 'JUP', 'PYTH', 'RAY', 'RNDR', 'HNT',
]);
exports.KNOWN_MEMECOINS = new Set([
    'BRETT', 'DEGEN', 'TOSHI', 'HIGHER', 'BASED', 'MOCHI', 'NORMIE',
    'KEYCAT', 'DOGINME', 'BENJI', 'SKI',
    'POPCAT', 'MEW', 'BOME', 'WEN', 'MYRO', 'SAMO', 'SLERF', 'PONKE',
    'BOOK', 'MOODENG', 'PNUT', 'GOAT', 'ACT', 'AI16Z', 'FARTCOIN', 'GRIFFAIN', 'PENGU', 'TRUMP',
]);
function isNoteworthyToken(symbol) {
    const upper = symbol.toUpperCase();
    return exports.MAJOR_TOKENS.has(upper) || exports.KNOWN_MEMECOINS.has(upper);
}
