// clude-bot/dist/utils/format.js
"use strict";
/**
 * Shared formatting utilities used across features.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.truncateWallet = truncateWallet;
exports.formatUsd = formatUsd;
exports.formatNumber = formatNumber;
exports.lamportsToSol = lamportsToSol;
exports.weiToEth = weiToEth;
exports.timeAgo = timeAgo;
exports.clamp = clamp;
exports.percentChange = percentChange;
function truncateWallet(address) {
    return `${address.slice(0, 4)}...${address.slice(-4)}`;
}
function formatUsd(amount) {
    if (amount >= 1_000_000_000)
        return `${(amount / 1_000_000_000).toFixed(1)}B`;
    if (amount >= 1_000_000)
        return `${(amount / 1_000_000).toFixed(1)}M`;
    if (amount >= 1_000)
        return `${(amount / 1_000).toFixed(0)}K`;
    return amount.toFixed(0);
}
function formatNumber(n) {
    if (n >= 1_000_000)
        return `${(n / 1_000_000).toFixed(1)}M`;
    if (n >= 1_000)
        return `${(n / 1_000).toFixed(0)}K`;
    return n.toString();
}
function lamportsToSol(lamports) {
    return lamports / 1e9;
}
function weiToEth(wei) {
    return Number(wei) / 1e18;
}
function timeAgo(dateStr) {
    const ms = Date.now() - new Date(dateStr).getTime();
    const hours = Math.floor(ms / (1000 * 60 * 60));
    if (hours < 1)
        return 'just now';
    if (hours < 24)
        return `${hours}h ago`;
    const days = Math.floor(hours / 24);
    if (days < 7)
        return `${days}d ago`;
    return `${Math.floor(days / 7)}w ago`;
}
function clamp(val, min, max) {
    return Math.max(min, Math.min(max, val));
}
function percentChange(oldVal, newVal) {
    if (oldVal === 0)
        return 0;
    return ((newVal - oldVal) / oldVal) * 100;
}
