// clude-bot/dist/utils/text.js
"use strict";
/**
 * Text processing utilities for mention handling, wallet extraction, and content analysis.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SOLANA_ADDRESS_REGEX = void 0;
exports.cleanMentionText = cleanMentionText;
exports.extractWalletAddress = extractWalletAddress;
exports.isQuestion = isQuestion;
exports.extractTokenMentions = extractTokenMentions;
exports.pickRandom = pickRandom;
exports.SOLANA_ADDRESS_REGEX = /[1-9A-HJ-NP-Za-km-z]{32,44}/g;
function cleanMentionText(text) {
    return text.replace(/@\w+/g, '').trim();
}
function extractWalletAddress(text) {
    const matches = text.match(exports.SOLANA_ADDRESS_REGEX);
    if (!matches)
        return null;
    return matches[0] || null;
}
function isQuestion(text) {
    const cleaned = text.replace(/@\w+/g, '').trim();
    return cleaned.includes('?') || /^(ask|what|why|how|when|where|will|should|can|is|do|does)\b/i.test(cleaned);
}
function extractTokenMentions(text) {
    const matches = text.match(/\$[A-Z]{2,10}/gi);
    return matches ? matches.map(t => t.toUpperCase()) : [];
}
function pickRandom(arr) {
    return arr[Math.floor(Math.random() * arr.length)];
}
