// clude-bot/dist/core/claude-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._setSystemPromptProvider = _setSystemPromptProvider;
exports._setResponsePostProcessor = _setResponsePostProcessor;
exports._setAnthropicClient = _setAnthropicClient;
exports.generateResponse = generateResponse;
exports.generateImportanceScore = generateImportanceScore;
exports.generateThread = generateThread;
const sdk_1 = __importDefault(require("@anthropic-ai/sdk"));
const config_1 = require("../config");
const logger_1 = require("./logger");
const log = (0, logger_1.createChildLogger)('claude-client');
// ── Pluggable system prompt & post-processor ──
// Defaults are generic SDK prompts. The bot wires in its personality at startup.
let _systemPromptProvider = () => 'You are an AI assistant with persistent memory.';
let _responsePostProcessor = (t) => t;
/** @internal Bot injects its personality (base-prompt.ts) here at startup. */
function _setSystemPromptProvider(fn) {
    _systemPromptProvider = fn;
}
/** @internal Bot injects its closer logic here at startup. */
function _setResponsePostProcessor(fn) {
    _responsePostProcessor = fn;
}
let anthropic;
try {
    anthropic = new sdk_1.default({ apiKey: config_1.config.anthropic.apiKey || 'placeholder' });
}
catch {
    anthropic = null;
}
let _anthropicOverride = null;
let _modelOverride = null;
function getClient() {
    return _anthropicOverride || anthropic;
}
function getModel() {
    return _modelOverride || config_1.config.anthropic.model;
}
/** @internal SDK escape hatch — allows Cortex to inject a pre-configured Anthropic client. */
function _setAnthropicClient(client, model) {
    _anthropicOverride = client;
    _modelOverride = model || null;
}
async function generateResponse(options) {
    const systemParts = [_systemPromptProvider()];
    if (options.memoryContext)
        systemParts.push(`\n\n${options.memoryContext}`);
    if (options.moodModifier)
        systemParts.push(`\n\n## Current Mood\n${options.moodModifier}`);
    if (options.tierModifier)
        systemParts.push(`\n\n## User Context\n${options.tierModifier}`);
    if (options.agentModifier)
        systemParts.push(`\n\n## Agent Context\n${options.agentModifier}`);
    if (options.featureInstruction)
        systemParts.push(`\n\n## Task\n${options.featureInstruction}`);
    const systemPrompt = systemParts.join('');
    // Build user content — ensure it's never empty
    let userContent = (options.userMessage || '').replace(/@\w+/g, '').trim();
    if (options.context) {
        userContent = `## Data\n${options.context}\n\n## User Message\n${userContent || '(no message, just a mention)'}`;
    }
    if (!userContent) {
        userContent = '(Someone mentioned you with no specific message. React to being summoned for nothing.)';
    }
    log.debug({ systemLength: systemPrompt.length, userLength: userContent.length }, 'Generating response');
    const response = await getClient().messages.create({
        model: getModel(),
        max_tokens: options.maxTokens || 300,
        temperature: 0.9,
        system: systemPrompt,
        messages: [{ role: 'user', content: userContent }],
    });
    let text = response.content
        .filter((block) => block.type === 'text')
        .map(block => block.text)
        .join('')
        .trim();
    // Strip any quotes the model may wrap the response in
    if (text.startsWith('"') && text.endsWith('"')) {
        text = text.slice(1, -1).trim();
    }
    // Apply post-processor (bot uses this for random closers)
    text = _responsePostProcessor(text);
    log.info({ responseLength: text.length }, 'Response generated');
    return text;
}
/**
 * Single-purpose LLM call to rate memory importance (Park et al. 2023).
 * Returns a raw string (expected: single integer 1-10).
 * Uses low maxTokens and temperature 0 for deterministic, fast scoring.
 */
async function generateImportanceScore(description) {
    const response = await getClient().messages.create({
        model: getModel(),
        max_tokens: 10,
        temperature: 0,
        system: 'You rate the importance of events for an AI agent called Clude. ' +
            'Clude is a tired, polite Solana meme token bot on X. ' +
            'Respond with ONLY a single integer from 1 to 10. ' +
            '1 = purely mundane (a greeting, a generic question). ' +
            '5 = moderately important (a returning user, a market opinion request). ' +
            '10 = extremely significant (a whale selling everything, a deeply personal interaction, an existential realization).',
        messages: [{
                role: 'user',
                content: `Rate the importance of this event for Clude:\n"${description.slice(0, 500)}"\nRating (1-10):`,
            }],
    });
    return response.content
        .filter((block) => block.type === 'text')
        .map(block => block.text)
        .join('')
        .trim();
}
async function generateThread(options) {
    const response = await generateResponse({
        ...options,
        maxTokens: 1200,
        featureInstruction: (options.featureInstruction || '') +
            '\n\nFormat: Write 3-5 tweets separated by ---. Each tweet must be under 270 characters.',
    });
    return response
        .split('---')
        .map(t => t.trim())
        .filter(t => t.length > 0)
        .map(t => t.slice(0, 280));
}
