// clude-bot/dist/core/embeddings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._configureEmbeddings = _configureEmbeddings;
exports.isEmbeddingEnabled = isEmbeddingEnabled;
exports.generateEmbedding = generateEmbedding;
exports.generateEmbeddings = generateEmbeddings;
const config_1 = require("../config");
const logger_1 = require("./logger");
const log = (0, logger_1.createChildLogger)('embeddings');
const PROVIDERS = {
    voyage: {
        url: 'https://api.voyageai.com/v1/embeddings',
        defaultModel: 'voyage-3-lite',
        authHeader: (key) => `Bearer ${key}`,
    },
    openai: {
        url: 'https://api.openai.com/v1/embeddings',
        defaultModel: 'text-embedding-3-small',
        authHeader: (key) => `Bearer ${key}`,
    },
};
let _enabled = null;
let _overrideConfig = null;
/** @internal SDK escape hatch — allows Cortex to override embedding config. */
function _configureEmbeddings(opts) {
    _overrideConfig = opts;
    _enabled = null; // reset cached check
}
function getEmbeddingConfig() {
    if (_overrideConfig)
        return _overrideConfig;
    return config_1.config.embedding;
}
function isEmbeddingEnabled() {
    if (_enabled !== null)
        return _enabled;
    const cfg = getEmbeddingConfig();
    _enabled = !!cfg.provider && !!cfg.apiKey && cfg.provider in PROVIDERS;
    if (_enabled) {
        log.info({ provider: cfg.provider }, 'Embedding system enabled');
    }
    return _enabled;
}
/**
 * Generate a single embedding vector for the given text.
 * Returns null if embeddings are disabled or the API call fails.
 */
async function generateEmbedding(text) {
    if (!isEmbeddingEnabled())
        return null;
    const cfg = getEmbeddingConfig();
    if (!cfg.provider || !(cfg.provider in PROVIDERS))
        return null;
    const providerConfig = PROVIDERS[cfg.provider];
    const model = cfg.model || providerConfig.defaultModel;
    try {
        const res = await fetch(providerConfig.url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': providerConfig.authHeader(cfg.apiKey),
            },
            body: JSON.stringify({
                input: text.slice(0, 8000),
                model,
                ...(cfg.provider === 'openai' ? { dimensions: cfg.dimensions } : {}),
            }),
        });
        if (!res.ok) {
            const errText = await res.text();
            log.error({ status: res.status, body: errText.slice(0, 200) }, 'Embedding API error');
            return null;
        }
        const data = await res.json();
        return data.data?.[0]?.embedding || null;
    }
    catch (err) {
        log.error({ err }, 'Embedding generation failed');
        return null;
    }
}
/**
 * Generate embeddings for multiple texts in a single batch API call.
 * More efficient than calling generateEmbedding() in a loop.
 * Returns an array matching input length; null for any that failed.
 */
async function generateEmbeddings(texts) {
    if (!isEmbeddingEnabled() || texts.length === 0)
        return texts.map(() => null);
    const cfg = getEmbeddingConfig();
    if (!cfg.provider || !(cfg.provider in PROVIDERS))
        return texts.map(() => null);
    const providerConfig = PROVIDERS[cfg.provider];
    const model = cfg.model || providerConfig.defaultModel;
    try {
        const res = await fetch(providerConfig.url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': providerConfig.authHeader(cfg.apiKey),
            },
            body: JSON.stringify({
                input: texts.map(t => t.slice(0, 8000)),
                model,
                ...(cfg.provider === 'openai' ? { dimensions: cfg.dimensions } : {}),
            }),
        });
        if (!res.ok) {
            log.error({ status: res.status }, 'Batch embedding API error');
            return texts.map(() => null);
        }
        const data = await res.json();
        const result = texts.map(() => null);
        for (const item of data.data || []) {
            result[item.index] = item.embedding;
        }
        return result;
    }
    catch (err) {
        log.error({ err }, 'Batch embedding generation failed');
        return texts.map(() => null);
    }
}
