// clude-bot/dist/core/inference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configureInference = configureInference;
exports.getInferenceConfig = getInferenceConfig;
exports.generate = generate;
exports.ask = ask;
const logger_1 = require("./logger");
const claude_client_1 = require("./claude-client");
const venice_client_1 = require("./venice-client");
const log = (0, logger_1.createChildLogger)('inference');
let inferenceConfig = {
    primary: 'auto',
    fallback: 'anthropic',
};
function configureInference(config) {
    inferenceConfig = { ...inferenceConfig, ...config };
    log.info({ primary: inferenceConfig.primary, fallback: inferenceConfig.fallback }, 'Inference configured');
}
function getInferenceConfig() {
    return inferenceConfig;
}
/**
 * Generate a response using the configured inference provider(s).
 * Handles automatic fallback if primary fails.
 */
async function generate(opts) {
    const provider = opts.provider || inferenceConfig.primary || 'auto';
    // Determine which providers to try
    const providers = getProviderOrder(provider);
    let lastError = null;
    for (const p of providers) {
        try {
            const result = await generateWithProvider(p, opts);
            if (result) {
                log.debug({ provider: p }, 'Inference succeeded');
                return result;
            }
        }
        catch (err) {
            lastError = err;
            log.warn({ provider: p, error: err.message }, 'Provider failed, trying next');
        }
    }
    throw lastError || new Error('All inference providers failed');
}
function getProviderOrder(provider) {
    if (provider === 'auto') {
        // Prefer Venice (decentralized) if available, fall back to Anthropic
        const providers = [];
        if ((0, venice_client_1.isVeniceEnabled)()) {
            providers.push('venice');
        }
        providers.push('anthropic');
        return providers;
    }
    const providers = [provider];
    if (inferenceConfig.fallback && inferenceConfig.fallback !== provider) {
        providers.push(inferenceConfig.fallback);
    }
    return providers;
}
async function generateWithProvider(provider, opts) {
    switch (provider) {
        case 'venice':
            return generateWithVenice(opts);
        case 'anthropic':
            return generateWithAnthropic(opts);
        default:
            throw new Error(`Unknown provider: ${provider}`);
    }
}
async function generateWithVenice(opts) {
    if (!(0, venice_client_1.isVeniceEnabled)()) {
        throw new Error('Venice not configured');
    }
    const messages = [];
    // Build system prompt
    let systemPrompt = opts.systemPrompt || 'You are Clude, an AI with persistent memory.';
    if (opts.featureInstruction) {
        systemPrompt += `\n\n${opts.featureInstruction}`;
    }
    // Add context if provided
    let userContent = opts.userMessage;
    if (opts.context) {
        userContent = `${opts.context}\n\n---\n\n${opts.userMessage}`;
    }
    messages.push({ role: 'user', content: userContent });
    return (0, venice_client_1.generateVeniceResponse)({
        messages,
        systemPrompt,
        maxTokens: opts.maxTokens,
        model: inferenceConfig.veniceModel,
    });
}
async function generateWithAnthropic(opts) {
    // Use existing Claude client
    return (0, claude_client_1.generateResponse)({
        userMessage: opts.userMessage,
        context: opts.context,
        featureInstruction: opts.featureInstruction,
        maxTokens: opts.maxTokens,
    });
}
/**
 * Quick helper for simple prompts.
 */
async function ask(prompt, opts) {
    return generate({ userMessage: prompt, ...opts });
}
