// clude-bot/dist/core/solana-client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConnection = getConnection;
exports._configureSolana = _configureSolana;
exports.getBotWallet = getBotWallet;
exports.writeMemo = writeMemo;
exports.writeMemoDevnet = writeMemoDevnet;
exports.verifySignature = verifySignature;
exports.solscanTxUrl = solscanTxUrl;
const web3_js_1 = require("@solana/web3.js");
const bs58Module = __importStar(require("bs58"));
const bs58 = bs58Module.default || bs58Module;
const config_1 = require("../config");
const logger_1 = require("./logger");
const tweetnacl_1 = __importDefault(require("tweetnacl"));
const constants_1 = require("../utils/constants");
const log = (0, logger_1.createChildLogger)('solana-client');
let connection;
let botWallet = null;
const MEMO_PROGRAM_ID = new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_ID);
function getConnection() {
    if (!connection) {
        connection = new web3_js_1.Connection(config_1.config.solana.rpcUrl, 'confirmed');
    }
    return connection;
}
/** @internal SDK escape hatch — allows Cortex to inject Solana config. */
function _configureSolana(rpcUrl, privateKey) {
    connection = new web3_js_1.Connection(rpcUrl, 'confirmed');
    if (privateKey) {
        try {
            const raw = privateKey.trim();
            let secretKey;
            if (raw.startsWith('[')) {
                secretKey = Uint8Array.from(JSON.parse(raw));
            }
            else {
                secretKey = bs58.decode(raw);
            }
            botWallet = web3_js_1.Keypair.fromSecretKey(secretKey);
        }
        catch (err) {
            log.error({ err }, 'SDK: Failed to load bot wallet');
        }
    }
}
function getBotWallet() {
    if (!botWallet && config_1.config.solana.botWalletPrivateKey) {
        try {
            let secretKey;
            const raw = config_1.config.solana.botWalletPrivateKey.trim();
            if (raw.startsWith('[')) {
                // JSON array format: [1,2,3,...]
                secretKey = Uint8Array.from(JSON.parse(raw));
            }
            else {
                // Base58 encoded format
                secretKey = bs58.decode(raw);
            }
            botWallet = web3_js_1.Keypair.fromSecretKey(secretKey);
            log.info({ publicKey: botWallet.publicKey.toBase58() }, 'Bot wallet loaded');
        }
        catch (err) {
            log.error({ err }, 'Failed to load bot wallet');
        }
    }
    return botWallet;
}
async function writeMemo(memo) {
    const wallet = getBotWallet();
    if (!wallet) {
        log.error('No bot wallet configured, cannot write memo');
        return null;
    }
    const conn = getConnection();
    const truncatedMemo = memo.slice(0, constants_1.MEMO_MAX_LENGTH);
    const instruction = new web3_js_1.TransactionInstruction({
        keys: [{ pubkey: wallet.publicKey, isSigner: true, isWritable: true }],
        programId: MEMO_PROGRAM_ID,
        data: Buffer.from(truncatedMemo, 'utf-8'),
    });
    const transaction = new web3_js_1.Transaction().add(instruction);
    try {
        const signature = await (0, web3_js_1.sendAndConfirmTransaction)(conn, transaction, [wallet]);
        log.info({ signature, memoLength: truncatedMemo.length }, 'Memo written on-chain');
        return signature;
    }
    catch (err) {
        log.error({ err }, 'Failed to write memo');
        return null;
    }
}
let devnetConnection;
function getDevnetConnection() {
    if (!devnetConnection) {
        devnetConnection = new web3_js_1.Connection('https://api.devnet.solana.com', 'confirmed');
    }
    return devnetConnection;
}
async function writeMemoDevnet(memo) {
    const wallet = getBotWallet();
    if (!wallet) {
        log.error('No bot wallet configured, cannot write devnet memo');
        return null;
    }
    const conn = getDevnetConnection();
    const truncatedMemo = memo.slice(0, constants_1.MEMO_MAX_LENGTH);
    const instruction = new web3_js_1.TransactionInstruction({
        keys: [{ pubkey: wallet.publicKey, isSigner: true, isWritable: true }],
        programId: MEMO_PROGRAM_ID,
        data: Buffer.from(truncatedMemo, 'utf-8'),
    });
    const transaction = new web3_js_1.Transaction().add(instruction);
    try {
        const signature = await (0, web3_js_1.sendAndConfirmTransaction)(conn, transaction, [wallet]);
        log.info({ signature, memoLength: truncatedMemo.length }, 'Memo written on devnet');
        return signature;
    }
    catch (err) {
        log.error({ err }, 'Failed to write devnet memo');
        return null;
    }
}
function verifySignature(message, signature, publicKey) {
    const messageBytes = new TextEncoder().encode(message);
    return tweetnacl_1.default.sign.detached.verify(messageBytes, signature, publicKey);
}
function solscanTxUrl(signature) {
    return `${constants_1.SOLSCAN_TX_BASE_URL}/${signature}`;
}
