// clude-bot/dist/core/venice-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VENICE_MODELS = void 0;
exports.initVenice = initVenice;
exports.isVeniceEnabled = isVeniceEnabled;
exports.getVeniceConfig = getVeniceConfig;
exports.generateVeniceResponse = generateVeniceResponse;
exports.generateVeniceResponseWithSearch = generateVeniceResponseWithSearch;
exports.askVenice = askVenice;
const logger_1 = require("./logger");
const log = (0, logger_1.createChildLogger)('venice');
// ============================================================
// VENICE CLIENT — Permissionless, Private LLM Inference
//
// Venice provides:
// - OpenAI-compatible API
// - No data retention (prompts stay private)
// - Permissionless access (no approval needed)
// - Multiple model access (Claude, GPT, open-source)
//
// This aligns with Clude's decentralization philosophy:
// - Memory: Solana (decentralized)
// - Inference: Venice (permissionless)
// ============================================================
const VENICE_API_URL = 'https://api.venice.ai/api/v1';
// Available Venice models (subset - they support many more)
exports.VENICE_MODELS = {
    // Reasoning
    'deepseek-r1': 'deepseek-ai-DeepSeek-R1',
    'qwen-reasoning': 'zai-org-glm-4.7-reasoning',
    // General
    'llama-70b': 'llama-3.3-70b',
    'qwen-235b': 'qwen3-235b-a22b',
    'glm-4': 'zai-org-glm-4.7',
    // Fast
    'llama-8b': 'llama-3.1-8b',
    'qwen-4b': 'qwen3-4b',
    // Uncensored
    'venice-uncensored': 'venice-uncensored-1.1',
};
let config = null;
function initVenice(cfg) {
    config = {
        ...cfg,
        model: cfg.model || exports.VENICE_MODELS['llama-70b'],
        maxTokens: cfg.maxTokens || 2000,
    };
    log.info({ model: config.model }, 'Venice client initialized');
}
function isVeniceEnabled() {
    return config !== null && !!config.apiKey;
}
function getVeniceConfig() {
    return config;
}
/**
 * Generate a response using Venice's permissionless LLM API.
 * OpenAI-compatible format.
 */
async function generateVeniceResponse(opts) {
    if (!config?.apiKey) {
        throw new Error('Venice client not initialized');
    }
    const messages = [];
    // Add system prompt if provided
    if (opts.systemPrompt) {
        messages.push({ role: 'system', content: opts.systemPrompt });
    }
    messages.push(...opts.messages);
    const model = opts.model || config.model || exports.VENICE_MODELS['llama-70b'];
    const maxTokens = opts.maxTokens || config.maxTokens || 2000;
    try {
        const response = await fetch(`${VENICE_API_URL}/chat/completions`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${config.apiKey}`,
            },
            body: JSON.stringify({
                model,
                messages,
                max_tokens: maxTokens,
                temperature: opts.temperature ?? 0.7,
                // Claude models on Venice require thinking config
                ...(model.startsWith('claude') && {
                    thinking: {
                        type: 'enabled',
                        budget_tokens: 2048,
                    },
                }),
            }),
        });
        if (!response.ok) {
            const errorText = await response.text();
            log.error({ status: response.status, error: errorText }, 'Venice API error');
            throw new Error(`Venice API error: ${response.status} ${errorText}`);
        }
        const data = (await response.json());
        if (!data.choices?.[0]?.message?.content) {
            throw new Error('Invalid Venice response format');
        }
        log.debug({
            model,
            promptTokens: data.usage?.prompt_tokens,
            completionTokens: data.usage?.completion_tokens,
        }, 'Venice response generated');
        return data.choices[0].message.content;
    }
    catch (err) {
        log.error({ err, model }, 'Venice request failed');
        throw err;
    }
}
/**
 * Generate a response with web search enabled.
 * Venice can search the web and include citations.
 */
async function generateVeniceResponseWithSearch(opts) {
    if (!config?.apiKey) {
        throw new Error('Venice client not initialized');
    }
    const messages = [];
    if (opts.systemPrompt) {
        messages.push({ role: 'system', content: opts.systemPrompt });
    }
    messages.push(...opts.messages);
    const model = opts.model || config.model || exports.VENICE_MODELS['llama-70b'];
    try {
        const response = await fetch(`${VENICE_API_URL}/chat/completions`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${config.apiKey}`,
            },
            body: JSON.stringify({
                model,
                messages,
                max_tokens: opts.maxTokens || 2000,
                venice_parameters: {
                    enable_web_search: 'on',
                    enable_web_citations: true,
                },
            }),
        });
        if (!response.ok) {
            const errorText = await response.text();
            throw new Error(`Venice API error: ${response.status} ${errorText}`);
        }
        const data = (await response.json());
        const content = data.choices[0]?.message?.content || '';
        // Extract citations if present (Venice includes them in the response)
        const citationMatches = content.match(/\[(\d+)\]\s*(https?:\/\/[^\s\]]+)/g);
        const citations = citationMatches?.map(m => m.replace(/\[\d+\]\s*/, ''));
        return { content, citations };
    }
    catch (err) {
        log.error({ err }, 'Venice web search request failed');
        throw err;
    }
}
/**
 * Simple helper for single-turn generation.
 */
async function askVenice(prompt, opts) {
    return generateVeniceResponse({
        messages: [{ role: 'user', content: prompt }],
        ...opts,
    });
}
